### DISCUSSION TITLE
Operators != and not-in

### DISCUSSION THREAD

**Discussion by Kaycell** [invertase/react-native-firebase#4399] [Q&A] [answered]
Howdy!

I'm pretty to the react / firebase community, but I've seen that we are missing operation string `!=` and `not-in` for firestore. I think they've been added recently, is there any plan to support them? I would be glad to contribute but I'm not sure of the amount of work necessary to support them.

**Comment by mikehardy** [comment, marked as answer]
Hi there! Someone else asked about this recently, the first thing to make sure of is that both firebase-android-sdk and firebase-ios-sdk have support - sometimes they add one before the other. In this case I believe they both have it now

Second thing is to find an analogous PR if possible, and with this one we are in luck as there was a PR recently that added a query modifier recently https://github.com/invertase/react-native-firebase/pull/3702/files

If you look through it's basically:
- modifying the javascript to understand the new query things
- adding the documentation in the types file
- implementing it in the two native parts
- testing

It honestly looked to me like the testing was the largest chunk of code, but having developed a feature in react-native-firebase before I can say you can't skip that step, as a test-driven development style is really the only way to go in this system.

To that end the testing module has a good document: https://github.com/invertase/react-native-firebase/blob/master/tests/README.md

When I did a feature recently I found it to be a multi-hour job vs multi-day but I was pretty focused and got the dev environment set up quickly and started with the e2e tests right away by creating a new e2e test and then limiting the test run to just that test, after which point I was moving quickly

The only warning I can give is that at this point we have a problem that I just noticed today and have not solved yet where the tests project is not compatible with Xcode 12 #4397 and until I fix that, it means developing features is either impossible or you have to install Xcode11 side by side. I will fix it quickly though as a blocker like that is not ok

Hope this helps, and I'd be happy to collaborate with any questions you have if you take a chance at it

**Reply by mikehardy** [reply]
Update: #4401 is nearly complete and should result in up to date SDKs and Xcode 12 being both usable and the default in the main branch, unblocking any activity here.

**Reply by mikehardy** [reply]
#4401 is merged and release, main branch and test project are open for development with up to date SDKs

**Reply by mikehardy** [reply]
Will be supported in #4474

**Reply by Kaycell** [reply]
Hi @mikehardy, thanks for the quick answer with a lot of details! Sadly I didn't have time to dive into the code and implement the feature. I'm glad someone took the time to do so! Do you know if this will be release any time soon? I don't know the release pace of the project. In the meantime is there any way to try it without too much conflict with other modules (auth, storage, etc..)?

**Reply by mikehardy** [reply]
@kleww 
It was released on Oct 30 as `@react-native-firebase/firestore` version 7.9.0 https://github.com/invertase/react-native-firebase/blob/master/packages/firestore/CHANGELOG.md#790-2020-10-30

If you use that version or later you'll have it

We release semi-frequently, depending on pace of contributions. Any time something gets merged, I try to release it. In general unless you really know what you are doing you should update all `@react-native-firebase` scoped packages at the same time, to the most current stable version. Be careful to read the CHANGELOG notes from https://rnfirebase.io/releases on each package that goes across a semver major boundary, as that indicates breaking changes and I leave notes in there on what you need to do to migrate
